### PR TITLE
"Don't directly throw Exception" System.Data.SqlClient

### LINKED ISSUE #25527: "Don't directly throw Exception" System.Data.SqlClient
I did some test [here](https://github.com/dotnet/corefx/compare/master...MarcoRossignoli:issue_23748_SqlClient#diff-4443802483f189746d4dcd5ca6e8d62c) on SqlClient(#23748) and i think it's a special case when exception is catch after throw. Maybe "InvalidOperationException"? @JonHanna do you think is fine here?
@saurabh500 @corivera

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
Closes #25527 and contributes to #23748

### CODE REVIEW

**Comment by MarcoRossignoli** [comment]
@dotnet-build-bot test Windows x86 Release Build

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs [hunk 1]
```diff
@@ -6339,7 +6339,7 @@ private void SNISSPIData(byte[] receivedBuff, UInt32 receivedLength, ref byte[]
                 {
                     _physicalStateObj.GenerateSspiClientContext(receivedBuff, receivedLength, ref sendBuff, ref sendLength, _sniSpnBuffer);
                 }
-                catch (Exception e)
+                catch (InvalidOperationException e)
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs]: Why did this change?  Previously it was catching everything and routing it through SSPIError; now it's only catching InvalidOperationExceptions.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs [hunk 2]
```diff
@@ -144,11 +144,11 @@ public void GenSspiClientContext(SspiClientContextStatus sspiClientContextStatus
                 // so we don't need to check for a GssApiException here.
                 if (statusCode.ErrorCode == SecurityStatusPalErrorCode.InternalError)
                 {
-                    throw new Exception(SQLMessage.KerberosTicketMissingError() + "\n" + statusCode);
+                    throw new InvalidOperationException(SQLMessage.KerberosTicketMissingError() + "\n" + statusCode);
                 }
                 else
                 {
-                    throw new Exception(SQLMessage.SSPIGenerateError() + "\n" + statusCode);
+                    throw new InvalidOperationException(SQLMessage.SSPIGenerateError() + "\n" + statusCode);
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: @saurabh500? @corivera?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs (hunk 1 above)
Comment (MarcoRossignoli) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs]: i agree with you, revert.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs (hunk 2 above)
Comment (MarcoRossignoli) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: on #25527 @saurabh500 said is "reasonable".

**Pull request merged**